frontend/lib/resolution/Resolver.cpp

```diff
@@ -1057,15 +1025,16 @@ QualifiedType Resolver::getTypeForDecl(const AstNode* declForErr,
                        QualifiedType(declKind, declaredType.type()));
     if (!got.passes()) {
       // For a record/union, check for an init= from the provided type
-      const Type* foundInitEqResultType =
-        tryResolveCrossTypeInitEq(declForErr, declaredType, initExprType);
-
-      if (!foundInitEqResultType) {
+      const bool isRecordOrUnion = (declaredType.type()->isRecordType() ||
+                                    declaredType.type()->isUnionType());
+      if (!(isRecordOrUnion &&
+            tryResolveInitEq(context, declForErr, declaredType.type(),
+                             initExprType.type(), poiScope))) {
         CHPL_REPORT(context, IncompatibleTypeAndInit, declForErr, typeForErr,
                     initForErr, declaredType.type(), initExprType.type());
         typePtr = ErroneousType::get(context);
       } else {
-        typePtr = foundInitEqResultType;
+        typePtr = declaredType.type();
```

Comment:
I don't think this is quite right, because we can have situations where an `init=` is used to create a type when the declared type is generic. I think it needs to get the type produced by the `init=` call.

For example, I think this example works/should work (I didn't try it in production, though):

``` chapel
var x: MyGenericRecord = 1;

record MyGenericRecord {
  type t;
  var x: t;
  proc init=(other: int) {
    this.t = int;
    this.x = other;
  }
}
```

Similar examples are even possible where it is more generic, as in `proc init=(other: numeric)`.

Reply:
I preserved the behavior of this from before the PR, but lost [this TODO](https://github.com/chapel-lang/chapel/pull/24058/files#diff-7e976ae78c2cdba0d88797590948e3655268da8fa82fdfb42cc18dbd2de526e4L958) that I think is referring to the same thing as you. For this PR, I think I should just add the TODO back, but at this point.